model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -125,24 +81,250 @@ impl Serialize for CommandDataOption {
     }
 }
 
-#[derive(Deserialize)]
-struct CommandDataOptionEnvelope {
-    name: String,
-    #[serde(rename = "type")]
-    kind: CommandOptionType,
-    #[serde(default)]
-    options: Vec<CommandDataOption>,
-    value: Option<CommandOptionValueEnvelope>,
-}
+impl<'de> Deserialize<'de> for CommandDataOption {
+    #[allow(clippy::too_many_lines)]
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Debug, Deserialize)]
+        #[serde(field_identifier, rename_all = "snake_case")]
+        enum Fields {
+            Name,
+            Type,
+            Options,
+            Value,
+        }
 
-#[derive(Debug, Deserialize)]
-#[serde(untagged)]
-enum CommandOptionValueEnvelope {
-    Boolean(bool),
-    Id(GenericId),
-    Integer(i64),
-    Number(Number),
-    String(String),
+        #[derive(Debug, Deserialize)]
+        #[serde(untagged)]
+        enum ValueEnvelope {
+            Boolean(bool),
+            Integer(i64),
+            Number(f64),
+            Id(GenericId),
+            String(String),
+        }
+
+        struct CommandDataOptionVisitor;
+
+        impl<'de> Visitor<'de> for CommandDataOptionVisitor {
+            type Value = CommandDataOption;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                formatter.write_str("CommandDataOption")
+            }
+
+            #[allow(clippy::too_many_lines)]
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::MapAccess<'de>,
+            {
+                let mut name_opt = None;
+                let mut kind_opt = None::<CommandOptionType>;
+                let mut options = Vec::<CommandDataOption>::new();
+                let mut value_opt = None::<ValueEnvelope>;
```

Comment:
Is the turbo fish syntax necessary? If yes could this follow code style (`let mut kind_opt: Option<CommandOptionType> = None;`)